Source/Particles/MultiParticleContainer.cpp

```diff
@@ -486,12 +486,19 @@ MultiParticleContainer::GetZeroChargeDensity (const int lev)
 {
     WarpX& warpx = WarpX::GetInstance();
 
-    BoxArray ba = warpx.boxArray(lev);
+    BoxArray nba = warpx.boxArray(lev);
     DistributionMapping dmap = warpx.DistributionMap(lev);
     const int ng_rho = warpx.get_ng_depos_rho().max();
 
-    auto zero_rho = std::make_unique<MultiFab>(amrex::convert(ba,IntVect::TheNodeVector()),
-                                               dmap,WarpX::ncomps,ng_rho);
+    bool is_PSATD_RZ = false;
+#ifdef WARPX_DIM_RZ
+    if (WarpX::maxwell_solver_id == MaxwellSolverAlgo::PSATD)
+        is_PSATD_RZ = true;
+#endif
+    if( !is_PSATD_RZ )
+        nba.surroundingNodes();
+
+    auto zero_rho = std::make_unique<MultiFab>(nba, dmap, WarpX::ncomps, ng_rho);
```

Comment:
The inconsistency between the box arrays of the zero rho and the filled rho in the RZ PSATD case was causing the test failures. I think making the zero array behave the same as the regular one is the right thing to do, but that needs someone more knowledgeable to confirm.

Reply:
ccing @RemiLehe and @EZoni for PSATD question.

Reply:
There's a broader or conceptually deeper inconsistency between RZ PSATD and everything else here, in that the diagnostic output will be slightly different depending on the solver used. At first, I was concerned that I'd have to fix *that* to get this working

Reply:
Hi, sorry I missed this. I can let @dpgrote and @RemiLehe comment as well, but I think rho should be always fully nodal (regardless of whether the simulation is nodal or staggered, which is controlled by `warpx.do_nodal`), except for RZ PSATD where all fields are fully cell-centered, including rho.

Now, what we do for the diagnostics is kind of a choice, so I guess it's debatable. But anyways I think I agree with the fact that the "zero" rho array should behave the same as the regular rho array.

Two other small things I noticed are:
- I see that we use `WarpX::ncomps`, which should be fine, but I think extracting the number of components directly from the rho MultiFab (by calling the method `nComp`), whenever possible, would be more robust.
- I see that `ng_rho` is set to `warpx.get_ng_depos_rho().max()` and I'm not sure why we need to take the `max` here, instead of simply taking the `IntVect` itself. I would be in favor of the latter, unless there is a specific reason why the `max` is needed here.

I'm not sure if this answer any of the questions, but I hope it helps. @dpgrote and @RemiLehe can comment more.

Reply:
I think that this change is the correct thing to do, to make the rho created here consistent with the one that would be created with `GetChargeDensity`. This makes sense since the will be used in the same contexts.

Note that here, there is no place to get `ncomps` from other than `Warpx::ncomps` since there is no multifab available to get the value from. This is consistent with `GetChargeDensity` which also uses `WarpX::ncomps`. Similarly, `GetChargeDensity` also uses `warpx.get_ng_depos_rho().max()`.

Reply:
Thinking about this some more, here is a suggestion for a later PR perhaps to reduce duplicate code. Since almost everything done in this routine is using the `warpx` instance, a better place for this might be in the `WarpX` class, perhaps a routine that returns a ready made rho MultiFab. Then the routine `WarpXParticleContainer::GetChargeDensity` could call that WarpX function to get the rho MultiFab instead of creating one itself.